File: 백준/Gold/15926. 현욱은 괄호왕이야！！/현욱은 괄호왕이야！！.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
let input = [];
let mx = 0;
let dp = new Array(200004).fill(0);

rl.on("line", (line) => {
  input.push(line);
}).on("close", () => {
  const n = Number(input[0]);
  const a = input.slice(1)[0].split("");
  let stk = [];
  let cnt = 0;
  for (let i = 0; i < n; i++) {
    if (a[i] === "(") stk.push(i);
    else {
      if (stk.length) {
        dp[i] = dp[stk.pop()] = 1;
      }
    }
  }
  for (let i = 0; i < n; i++) {
    if (dp[i]) {
      cnt++;
      mx = Math.max(mx, cnt);
    } else cnt = 0;
  }
  console.log(mx);
});

File: 백준/Gold/15926. 현욱은 괄호왕이야！！/현욱은 괄호왕이야！！.cc
#include <bits/stdc++.h>
using namespace std;
int n, d[200004], cnt, mx;
string s;
int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  stack<int> stk;
  cin >> n >> s;
  for (int i = 0; i < s.size(); i++)
  {
    if (s[i] == '(')
      stk.push(i);
    else if (stk.size())
    {
      d[i] = d[stk.top()] = 1;
      stk.pop();
    }
  }
  for (int i = 0; i < s.size(); i++)
  {
    if (d[i])
    {
      cnt++;
      mx = max(mx, cnt);
    }
    else
      cnt = 0;
  }
  cout << mx << '\n';
  return 0;
}